bdd/spec/011_array_spec.sh

```diff
@@ -373,27 +373,37 @@ true"
     End
   End
 
-  Describe "everything else ;)"
+  Describe "reduce, filter, and concat"
     before() {
-      # TODO: sourceToAll
-      sourceToTemp "
+      sourceToAll "
         from @std/app import start, print, exit
 
-        fn isOdd(val: int64): bool {
-          return val % 2 == 1
-        }
-
         on start {
           const test = [ 1, 1, 2, 3, 5, 8 ]
+          const test2 = [ 4, 5, 6 ]
           print('reduce test')
-          test.reduce(add).print()
+          test.reduce(fn (a: int, b: int): int = a + b).print()
 
           print('filter test')
-          test.filter(isOdd).each(print)
+          test.filter(fn (val: int64): bool {
+            return val % 2 == 1
+          }).map(fn (val: int64): string {
+            return toString(val)
+          }).join(', ').print()
 
           print('concat test')
-          test.concat(test2).each(print)
-          (test + test2).each(print)
+          test.concat(test2).map(fn (val: int64): string {
+            return toString(val)
+          }).join(', ').print()
+          (test + test2).map(fn (val: int64): string {
+            return toString(val)
+          }).join(', ').print()
+
+          print('reduce as filter and concat test')
+          // TODO: Lots of improvements needed for closures passed directly to opcodes. This one-liner is ridiculous
```

Comment:
So, closures passed directly to opcodes cannot use any opcodes themselves that are IO-bound, or trigger opcode stack changes like `condfn` does. That means conditionals are out, but also something like `urls.map(fn (url: string) = http.get(url))` will *also* not work.

Then there's the issue with return types being required for these closures but not required anywhere else (including closures passed in to user-defined functions) which makes for weird error messages.

Then there's the issue of closures being converted to microstatements *too early* so the microstatement types may be wrong, making it nearly impossible to write a closure used by multiple opcodes.

Then there's the issue of closure selection not occurring correctly (it just picks the first match by name if passed in by name instead of inline declared) so you can't do something like `arr.each(print)` because it will choose the "wrong" print function and fail.

Then there's the issue of not being able to use an opcode itself as your closure function, so `const sum = arr.reduce(add)` doesn't work.

"Lots of improvements". Each of these should probably be turned into a Github Issue, but I don't want all of that in the code comments. :/

Reply:
Wow, got it. It is good that we have it in this PR at least.